### ISSUE TITLE
[jvm-packages] xgboost4j-spark XGBoostClassifier cannot assign 0 to max_depth

### ISSUE THREAD

**Issue by SpringHerald** [dmlc/xgboost#4038]
The xgboost spark API said "maximum depth of a tree, increase this value will make model more complex / likely to be overfitting. [default=6] range: [1, Int.MaxValue]".
But [xgboost4j-spark tutorial](https://xgboost.readthedocs.io/en/latest/jvm/xgboost4j_spark_tutorial.html#training) said "In XGBoost4J-Spark, we support not only the default set of parameters but also the camel-case variant of these parameters to keep consistent with Spark’s MLLIB parameters."
And [XGBoost Parameters](https://xgboost.readthedocs.io/en/latest/parameter.html) notified that max_depth=0 indicates no limit.
So maybe it's better to enable max_depth=0 in xgboost4j-spark?

```
Exception in thread "main" java.lang.IllegalArgumentException: xgbc_8554b906dd6c parameter maxDepth given invalid value 0.
	at org.apache.spark.ml.param.Param.validate(params.scala:77)
	at org.apache.spark.ml.param.ParamPair.<init>(params.scala:528)
	at org.apache.spark.ml.param.Param.$minus$greater(params.scala:87)
	at org.apache.spark.ml.param.Params$class.set(params.scala:609)
	at org.apache.spark.ml.PipelineStage.set(Pipeline.scala:42)
	at org.apache.spark.ml.param.Params$class.set(params.scala:616)
	at org.apache.spark.ml.PipelineStage.set(Pipeline.scala:42)
	at ml.dmlc.xgboost4j.scala.spark.params.ParamMapFuncs$$anonfun$XGBoostToMLlibParams$2.apply(GeneralParams.scala:240)
	at ml.dmlc.xgboost4j.scala.spark.params.ParamMapFuncs$$anonfun$XGBoostToMLlibParams$2.apply(GeneralParams.scala:225)
	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
	at scala.collection.immutable.Map$Map4.foreach(Map.scala:188)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
	at ml.dmlc.xgboost4j.scala.spark.params.ParamMapFuncs$class.XGBoostToMLlibParams(GeneralParams.scala:225)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.XGBoostToMLlibParams(XGBoostClassifier.scala:48)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.<init>(XGBoostClassifier.scala:61)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.<init>(XGBoostClassifier.scala:58)
	at cn.com.bsfit.dm.components.algo.XGBoostTrial$.main(XGBoostTrial.scala:40)
	at cn.com.bsfit.dm.components.algo.XGBoostTrial.main(XGBoostTrial.scala)
```

**Comment by CodingCat** [comment]
0 is not a valid value for the current growing policy adopted in spark case, though it would be valid if you use fast histogram in future https://github.com/dmlc/xgboost/pull/4011

@hcho3 any insights on why we said 0 means no limit in the doc? I think it's only for fast histogram?

**Comment by CodingCat** [comment]
in future, the usage discussion should be to https://discuss.xgboost.ai/

**Comment by hcho3** [comment]
@CodingCat The doc says

> Note that limit is required when grow_policy is set of depthwise

**Comment by CodingCat** [comment]
isn't this incorrect? https://github.com/dmlc/xgboost/blob/master/src/tree/updater_histmaker.cc#L136

for approx, setting to zero will have an empty tree....

this no limit thing looks like only applied to loss-guide growing in fast histogram

**Comment by hcho3** [comment]
Yes, it looks like the no-limit condition is only applicable to the fast histogram.

**Comment by rongou** [comment]
You can also use no-limit and `lossguide` with `gpu_hist`, which I'm trying to add support for. Would it be possible to relax the check to `>=0`?
